File: $_algotsu-kaizen/problemo-set/number-complement.js
/**
 * 476. Number Complement
 * https://leetcode.com/problems/number-complement/description/
 */

/**
 * @param {number} num
 * @return {number}
 */
var findComplement = function (num) {
    const bit = new Uint32Array(0x1)///0b1
    bit[0x0] = 0b1
    const U_INT_0x20 = new Uint32Array(0x1)
    U_INT_0x20[0x0] = Number(num)

    while (bit <= U_INT_0x20[0x0]) {
        if (U_INT_0x20[0x0] & bit[0])
            U_INT_0x20[0x0] ^= bit[0]
        else
            U_INT_0x20[0x0] |= bit[0]
        bit[0] = bit[0] << 0b1
    }

    return U_INT_0x20[0x0]
};

export { findComplement }

// Input: nums = 5
// Output: 2
console.log("$#",
    findComplement(
        5
    ),
    2
);

// Input: nums = 1
// Output: 0
console.log("$#",
    findComplement(
        1
    ),
    0
);

// Input: nums = 2147483647
// Output: 0
console.log("$#",
    findComplement(
        2147483647
    ),
    0
);

File: $_algotsu-kaizen/problemo-set/number-complement.rs
struct Solution;

impl Solution {
    pub fn find_complement(num: i32) -> i32 {
        let mut bit: u32 = 1;
        let mut U_INT_0x20: u32 = num as u32;

        while bit <= U_INT_0x20 {
            if U_INT_0x20 & bit != 0 {
                U_INT_0x20 ^= bit;
            } else {
                U_INT_0x20 |= bit;
            }
            bit = bit << 0b1;
        }

        return U_INT_0x20 as i32;
    }
}

fn main() {
    println!("The {}th ugly number is: {}", 10, Solution::find_complement(5));
    println!("The {}th ugly number is: {}", 1, Solution::find_complement(1));
    println!("The {}th ugly number is: {}", 1, Solution::find_complement(2147483647));
}
